File: app/src/main/java/com/example/energo_monitoring/data/db/ResultDataDAO.java
package com.example.energo_monitoring.data.db;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;

import com.example.energo_monitoring.data.FlowTransducerCheckLengthResult;
import com.example.energo_monitoring.data.api.ClientInfo;
import com.example.energo_monitoring.data.api.OrganizationInfo;
import com.example.energo_monitoring.data.api.ProjectDescription;
import com.example.energo_monitoring.data.devices.DeviceCounter;
import com.example.energo_monitoring.data.devices.DeviceFlowTransducer;
import com.example.energo_monitoring.data.devices.DevicePressureTransducer;
import com.example.energo_monitoring.data.devices.DeviceTemperatureCounter;
import com.example.energo_monitoring.data.devices.DeviceTemperatureTransducer;

import java.util.List;

@Dao
public abstract class ResultDataDAO {
//    @Insert(onConflict = OnConflictStrategy.REPLACE)
//    public abstract void insertData(ClientInfo clientInfo);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertOrganizationInfo(OrganizationInfo organizationInfo);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertProjectDescription(ProjectDescription projectDescription);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertClientInfo(ClientInfo clientInfo);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertOtherInfo(OtherInfo otherInfo);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertDeviceTemperatureCounter(List<DeviceTemperatureCounter> devices);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertDeviceFlowTransducers(List<DeviceFlowTransducer> devices);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertDeviceTemperatureTransducers(List<DeviceTemperatureTransducer> devices);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertDevicePressureTransducers(List<DevicePressureTransducer> devices);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertDeviceCounters(List<DeviceCounter> devices);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insertFlowTransducerCheckLengthResults(List<FlowTransducerCheckLengthResult> results);

    @Query("DELETE FROM OtherInfo WHERE dataId = :dataId")
    public abstract void deleteOtherInfo(int dataId);

    @Query("DELETE FROM ProjectDescription WHERE dataId = :dataId")
    public abstract void deleteProjectDescription(int dataId);

    @Query("DELETE FROM DeviceTemperatureCounter WHERE dataId = :dataId")
    public abstract void deleteDeviceTemperatureCounter(int dataId);

    @Query("DELETE FROM DeviceFlowTransducer WHERE dataId = :dataId")
    public abstract void deleteDeviceFlowTransducers(int dataId);

    @Query("DELETE FROM DeviceTemperatureTransducer WHERE dataId = :dataId")
    public abstract void deleteDeviceTemperatureTransducers(int dataId);

    @Query("DELETE FROM DevicePressureTransducer WHERE dataId = :dataId")
    public abstract void deleteDevicePressureTransducers(int dataId);

    @Query("DELETE FROM DeviceCounter WHERE dataId = :dataId")
    public abstract void deleteDeviceCounters(int dataId);

    @Query("DELETE FROM Clientinfo WHERE dataId = :dataId")
    public abstract void deleteClientInfo(int dataId);

    @Query("SELECT * FROM ProjectDescription WHERE dataId = :dataId")
    public abstract ProjectDescription getProjectDescription(int dataId);

    @Query("SELECT * FROM DevicePressureTransducer WHERE dataId = :dataId")
    public abstract List<DeviceTemperatureCounter> getDeviceTemperatureCounter(int dataId);

    @Query("SELECT * FROM DeviceFlowTransducer WHERE dataId = :dataId")
    public abstract List<DeviceFlowTransducer> getDeviceFlowTransducers(int dataId);

    @Query("SELECT * FROM DeviceTemperatureTransducer WHERE dataId = :dataId")
    public abstract List<DeviceTemperatureTransducer> getDeviceTemperatureTransducers(int dataId);

    @Query("SELECT * FROM DevicePressureTransducer WHERE dataId = :dataId")
    public abstract List<DevicePressureTransducer> getDevicePressureTransducers(int dataId);

    @Query("SELECT * FROM OrganizationInfo WHERE dataId = :dataId")
    public abstract OrganizationInfo getOrganizationInfo(int dataId);

    @Query("SELECT * FROM ClientInfo WHERE dataId = :dataId")
    public abstract ClientInfo getClientInfo(int dataId);

    @Query("SELECT * FROM OtherInfo WHERE dataId = :dataId")
    public abstract OtherInfo getOtherInfo(int dataId);

    @Transaction
    @Query("SELECT * FROM OtherInfo WHERE dataId = :id")
    public abstract ResultData getData(int id);
}

File: app/src/main/java/com/example/energo_monitoring/checks/data/db/ResultDataDAO.kt
package com.example.energo_monitoring.checks.data.db

import androidx.room.*
import com.example.energo_monitoring.checks.data.api.ProjectDescription
import com.example.energo_monitoring.checks.data.api.OrganizationInfo
import com.example.energo_monitoring.checks.data.api.ClientInfo
import com.example.energo_monitoring.checks.data.devices.DeviceTemperatureCounter
import com.example.energo_monitoring.checks.data.devices.DeviceFlowTransducer
import com.example.energo_monitoring.checks.data.devices.DeviceTemperatureTransducer
import com.example.energo_monitoring.checks.data.devices.DevicePressureTransducer
import com.example.energo_monitoring.checks.data.devices.DeviceCounter
import com.example.energo_monitoring.checks.data.FlowTransducerLength
import com.example.energo_monitoring.checks.data.ProjectFile

@Dao
abstract class ResultDataDAO {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertOrganizationInfo(organizationInfo: OrganizationInfo)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertProjectDescription(projectDescription: ProjectDescription)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertClientInfo(clientInfo: ClientInfo)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertOtherInfo(otherInfo: OtherInfo)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertDeviceTemperatureCounters(devices: List<DeviceTemperatureCounter>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertDeviceFlowTransducers(devices: List<DeviceFlowTransducer?>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertDeviceTemperatureTransducers(devices: List<DeviceTemperatureTransducer?>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertDevicePressureTransducers(devices: List<DevicePressureTransducer?>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertDeviceCounters(devices: List<DeviceCounter?>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertFlowTransducerCheckLengthResults(results: List<FlowTransducerLength?>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertProjectFiles(files: List<ProjectFile?>)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    abstract fun insertProjectFile(file: ProjectFile)

    @Query("DELETE FROM OtherInfo WHERE dataId = :dataId")
    abstract fun deleteOtherInfo(dataId: Int)

    @Query("DELETE FROM ProjectDescription WHERE dataId = :dataId")
    abstract fun deleteProjectDescription(dataId: Int)

    @Query("DELETE FROM DeviceTemperatureCounter WHERE dataId = :dataId")
    abstract fun deleteDeviceTemperatureCounter(dataId: Int)

    @Query("DELETE FROM DeviceFlowTransducer WHERE dataId = :dataId")
    abstract fun deleteDeviceFlowTransducers(dataId: Int)

    @Query("DELETE FROM DeviceTemperatureTransducer WHERE dataId = :dataId")
    abstract fun deleteDeviceTemperatureTransducers(dataId: Int)

    @Query("DELETE FROM DevicePressureTransducer WHERE dataId = :dataId")
    abstract fun deleteDevicePressureTransducers(dataId: Int)

    @Query("DELETE FROM DeviceCounter WHERE dataId = :dataId")
    abstract fun deleteDeviceCounters(dataId: Int)

    @Query("DELETE FROM Clientinfo WHERE dataId = :dataId")
    abstract fun deleteClientInfo(dataId: Int)

    @Query("SELECT * FROM ProjectDescription WHERE dataId = :dataId")
    abstract fun getProjectDescription(dataId: Int): ProjectDescription?

    @Query("SELECT * FROM DeviceCounter WHERE dataId = :dataId")
    abstract fun getDeviceCounters(dataId: Int): List<DeviceCounter>?

    @Query("SELECT * FROM DeviceTemperatureCounter WHERE dataId = :dataId")
    abstract fun getDeviceTemperatureCounter(dataId: Int): List<DeviceTemperatureCounter>?

    @Query("SELECT * FROM DeviceFlowTransducer WHERE dataId = :dataId")
    abstract fun getDeviceFlowTransducers(dataId: Int): List<DeviceFlowTransducer>?

    @Query("SELECT * FROM DeviceTemperatureTransducer WHERE dataId = :dataId")
    abstract fun getDeviceTemperatureTransducers(dataId: Int): List<DeviceTemperatureTransducer>?

    @Query("SELECT * FROM DevicePressureTransducer WHERE dataId = :dataId")
    abstract fun getDevicePressureTransducers(dataId: Int): List<DevicePressureTransducer>?

    @Query("SELECT * FROM OrganizationInfo WHERE dataId = :dataId")
    abstract fun getOrganizationInfo(dataId: Int): OrganizationInfo?

    @Query("SELECT * FROM ClientInfo WHERE dataId = :dataId")
    abstract fun getClientInfo(dataId: Int): ClientInfo?

    @Query("SELECT * FROM OtherInfo WHERE dataId = :dataId")
    abstract fun getOtherInfo(dataId: Int): OtherInfo?

    @Query("SELECT * FROM ProjectFile WHERE dataId = :dataId")
    abstract fun getProjectFiles(dataId: Int): List<ProjectFile?>?

    @Transaction
    @Query("SELECT * FROM OtherInfo WHERE dataId = :id")
    abstract fun getData(id: Int): ResultData?
}